### ISSUE TITLE
About zero initialization

### ISSUE THREAD

**Issue by imabackstabber** [yz93/LAVT-RIS#25]
Hi, I've been reading your article lately, you guys mentioned that apply zero initialization to gate, which will ease the start of training.
In code, it's written as:
```python
          self.res_gate = nn.Sequential(
                  nn.Linear(dim, dim, bias=False),
                  nn.ReLU(),
                  nn.Linear(dim, dim, bias=False),
                  nn.Tanh()
              )

        # ...

        # initialize the gate to 0
        nn.init.zeros_(self.res_gate[0].weight)
        nn.init.zeros_(self.res_gate[2].weight)

    def forward(self, x, H, W, l, l_mask):
        """ Forward function.

        Args:
            x: Input feature, tensor size (B, H*W, C).
            H, W: Spatial resolution of the input feature.
        """

        # ...

        # PWAM fusion
        x_residual = self.fusion(x, l, l_mask)
        # apply a gate on the residual
        x = x + (self.res_gate(x_residual) * x_residual)
```

I wonder why this could work, cause it seems that grad of `self.res_gate` will always be 0? This is my toy script implementing that(basicly):
```python
import torch
import torch.nn as nn

a_1= nn.Sequential(
    nn.Linear(10,10,bias=False),
    nn.ReLU()
)
a_2= nn.Sequential(
    nn.Linear(10,10,bias=False),
    nn.Tanh()
)

b = nn.Linear(10,10, bias= False)

fake = torch.randn((2,10))


nn.init.zeros_(a_1[0].weight)
nn.init.zeros_(a_2[0].weight)

a_1_optim = torch.optim.Adam(a_1.parameters())
a_2_optim = torch.optim.Adam(a_2.parameters())
b_optim = torch.optim.Adam(b.parameters())


for i in range(2):
    intermediate = a_1(fake)
    output = (a_2(intermediate) * fake) + b(fake)

    loss = output.sum()
    a_1_optim.zero_grad()
    a_2_optim.zero_grad()
    b_optim.zero_grad()
    loss.backward()

    # print('a_1', a_1[0].weight)
    # print('intermediate:', intermediate)
    # print('a_2',a_2[0].weight)
    # print(b.weight.grad)
    # print(b.weight)
    # print('before:', a_1[0].weight)
    print("a_1[0] grad:", a_1[0].weight.grad)
    print("a_2[0] grad:", a_2[0].weight.grad)
    # print('but a[2] contains:', a_1[0].weight.grad)

    a_1_optim.step()
    a_2_optim.step()
    b_optim.step()

```
Above script will output result below:
```bash
(mic) alex@path % python check.py
a_1[0] grad: tensor([[0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]])
a_2[0] grad: tensor([[0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]])
a_1[0] grad: tensor([[0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]])
a_2[0] grad: tensor([[0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]])
```
so it seems that, at least in this script, linear layers will always be 0?
My env:
```bash
MacBook Air (M1, 2020)
torch: 1.12.1
```
I'll be more than thankful if you could reply this issue promptly.

**Comment by yz93** [comment]
Hi, thanks!

"you guys mentioned that apply zero initialization to gate, which will ease the start of training"---I don't remember that we claimed this in our paper. Where did you find it in the paper? I might have overlooked. 


It does seem that these two lines of code have some problems.

---
        # initialize the gate to 0
        nn.init.zeros_(self.res_gate[0].weight)
        nn.init.zeros_(self.res_gate[2].weight)
---

Could you in your script try to only zero-initialize the second linear layer while randomly initializing the first linear layer, and see if all gradients are still zeros?

Btw, someone mentioned in an earlier issue, that those two lines of code did not have an actual effect because some other code did random initialization later on, which overwrote the zero weights. I should check that again when I get the time and update the code.

**Comment by imabackstabber** [comment]
I canceled out the first zero initialization, it seems that it's getting right:
```python
# nn.init.zeros_(a_1[0].weight)
nn.init.zeros_(a_2[0].weight)
```
it outputs:
```bash
(mic) alex@ALEXWYANG-MB0 python_playground % python check.py
a_1[0] grad: tensor([[0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]])
a_2[0] grad: tensor([[ 1.8060e-01,  5.4063e-01,  3.2481e-01,  0.0000e+00,  8.6390e-02,
          3.2595e-01,  0.0000e+00,  6.6718e-01,  2.5977e-01,  2.1201e-01],
        [ 4.0218e-01,  8.2015e-01,  8.7780e-01,  0.0000e+00,  2.3347e-01,
          3.6120e-01,  0.0000e+00,  1.0259e+00,  7.0202e-01,  5.0067e-01],
        [ 2.4444e-01,  2.0872e+00, -1.0591e-01,  0.0000e+00, -2.8169e-02,
          1.7291e+00,  0.0000e+00,  2.5272e+00, -8.4700e-02,  1.8613e-01],
        [-4.2558e-01, -5.6306e-01, -1.0515e+00,  0.0000e+00, -2.7968e-01,
         -9.2592e-02,  0.0000e+00, -7.2034e-01, -8.4098e-01, -5.5248e-01],
        [-3.1787e-01, -7.3973e-01, -6.5697e-01,  0.0000e+00, -1.7473e-01,
         -3.7243e-01,  0.0000e+00, -9.2048e-01, -5.2541e-01, -3.8892e-01],
        [-6.9308e-01, -2.5710e+00, -1.0468e+00,  0.0000e+00, -2.7842e-01,
         -1.7224e+00,  0.0000e+00, -3.1550e+00, -8.3719e-01, -7.7671e-01],
        [-1.8251e-01, -1.6153e+00,  1.0199e-01,  0.0000e+00,  2.7126e-02,
         -1.3451e+00,  0.0000e+00, -1.9551e+00,  8.1565e-02, -1.3474e-01],
        [-2.5470e-01, -2.7791e-01, -6.5311e-01,  0.0000e+00, -1.7371e-01,
          7.0861e-04,  0.0000e+00, -3.6033e-01, -5.2232e-01, -3.3504e-01],
        [ 5.3439e-02, -1.3477e+00,  7.0292e-01,  0.0000e+00,  1.8696e-01,
         -1.3361e+00,  0.0000e+00, -1.6091e+00,  5.6216e-01,  1.7488e-01],
        [ 2.2954e-02,  7.4170e-01, -2.2958e-01,  0.0000e+00, -6.1062e-02,
          6.8089e-01,  0.0000e+00,  8.9119e-01, -1.8361e-01, -2.3114e-02]])
a_1[0] grad: tensor([[-4.5178e-03, -9.4015e-03, -8.4435e-03,  9.4248e-03,  7.5879e-03,
          1.8190e-02,  6.4019e-03,  5.5391e-03,  1.2532e-03, -1.7303e-03],
        [-4.6122e-03, -8.2390e-03, -1.3419e-02,  7.1045e-03,  6.9964e-03,
          2.0327e-02,  1.0321e-02,  3.9392e-03,  6.6177e-03, -4.1494e-03],
        [-3.2513e-03, -8.7866e-03,  1.0601e-03,  1.0526e-02,  6.5761e-03,
          1.0478e-02, -1.0209e-03,  6.5375e-03, -7.0361e-03,  2.2981e-03],
        [ 0.0000e+00,  0.0000e+00,  0.0000e+00,  0.0000e+00,  0.0000e+00,
          0.0000e+00,  0.0000e+00,  0.0000e+00,  0.0000e+00,  0.0000e+00],
        [-3.2513e-03, -8.7866e-03,  1.0601e-03,  1.0526e-02,  6.5761e-03,
          1.0478e-02, -1.0209e-03,  6.5375e-03, -7.0361e-03,  2.2981e-03],
        [-2.6507e-03, -2.9373e-03, -1.4061e-02,  7.5298e-04,  3.0287e-03,
          1.4007e-02,  1.0939e-02, -5.7626e-06,  1.0865e-02, -5.5371e-03],
        [ 0.0000e+00,  0.0000e+00,  0.0000e+00,  0.0000e+00,  0.0000e+00,
          0.0000e+00,  0.0000e+00,  0.0000e+00,  0.0000e+00,  0.0000e+00],
        [-4.6122e-03, -8.2390e-03, -1.3419e-02,  7.1045e-03,  6.9964e-03,
          2.0327e-02,  1.0321e-02,  3.9392e-03,  6.6177e-03, -4.1494e-03],
        [-3.2513e-03, -8.7866e-03,  1.0601e-03,  1.0526e-02,  6.5761e-03,
          1.0478e-02, -1.0209e-03,  6.5375e-03, -7.0361e-03,  2.2981e-03],
        [-4.3242e-03, -9.6008e-03, -5.9552e-03,  1.0136e-02,  7.5952e-03,
          1.6633e-02,  4.4506e-03,  6.0620e-03, -1.1659e-03, -6.0035e-04]])
a_2[0] grad: tensor([[ 1.8060e-01,  5.4062e-01,  3.2481e-01,  0.0000e+00,  8.6389e-02,
          3.2595e-01,  0.0000e+00,  6.6717e-01,  2.5976e-01,  2.1201e-01],
        [ 4.0217e-01,  8.2014e-01,  8.7779e-01,  0.0000e+00,  2.3347e-01,
          3.6119e-01,  0.0000e+00,  1.0259e+00,  7.0202e-01,  5.0067e-01],
        [ 2.4444e-01,  2.0872e+00, -1.0591e-01,  0.0000e+00, -2.8169e-02,
          1.7291e+00,  0.0000e+00,  2.5271e+00, -8.4700e-02,  1.8613e-01],
        [-4.2557e-01, -5.6305e-01, -1.0515e+00,  0.0000e+00, -2.7968e-01,
         -9.2590e-02,  0.0000e+00, -7.2033e-01, -8.4097e-01, -5.5247e-01],
        [-3.1787e-01, -7.3972e-01, -6.5696e-01,  0.0000e+00, -1.7473e-01,
         -3.7242e-01,  0.0000e+00, -9.2046e-01, -5.2541e-01, -3.8891e-01],
        [-6.9307e-01, -2.5709e+00, -1.0468e+00,  0.0000e+00, -2.7842e-01,
         -1.7223e+00,  0.0000e+00, -3.1549e+00, -8.3719e-01, -7.7670e-01],
        [-1.8251e-01, -1.6153e+00,  1.0199e-01,  0.0000e+00,  2.7126e-02,
         -1.3451e+00,  0.0000e+00, -1.9551e+00,  8.1565e-02, -1.3473e-01],
        [-2.5470e-01, -2.7791e-01, -6.5310e-01,  0.0000e+00, -1.7371e-01,
          7.0861e-04,  0.0000e+00, -3.6033e-01, -5.2232e-01, -3.3504e-01],
        [ 5.3441e-02, -1.3477e+00,  7.0292e-01,  0.0000e+00,  1.8696e-01,
         -1.3361e+00,  0.0000e+00, -1.6091e+00,  5.6216e-01,  1.7488e-01],
        [ 2.2953e-02,  7.4169e-01, -2.2958e-01,  0.0000e+00, -6.1062e-02,
          6.8088e-01,  0.0000e+00,  8.9118e-01, -1.8361e-01, -2.3116e-02]])
```
seems fine with me.

> someone mentioned in an earlier issue, that those two lines of code did not have an actual effect because some other code did random initialization later on, which overwrote the zero weights. 

so seems that the only thing one should do is randomize the initial value. In my experiment, I noticed the forementioned 0 value and change to initialize weights with some relatively small number. Hope that will ease the training as hope so.
Btw, if you guys can confirm that a subsequent randomization does exist, I'll be much obliged if a follow-up comment can be added in this issue.

**Comment by yz93** [comment]
Hi, no problem.

I can confirm that the weights are randomly initialized.

[This line](https://github.com/yz93/LAVT-RIS/blob/0d1d9dc5a550ddea32100bcef8c2529656a84816/lib/backbone.py#L446) from the init_weights function did that. The init_weights function is called in segmentation.py (in lines 54, 57, 123, and 126), and when it is called, the initialization done by the constructor of MMBasicLayer is overwritten.

Thank you!

**Comment by imabackstabber** [comment]
ok, I'll close the issue then.
